core/src/main/java/org/apache/accumulo/core/summary/SummarySerializer.java

```diff
@@ -0,0 +1,543 @@
+      count++;
+
+      if (!sawFirst) {
+        firstRow = k.getRow();
+        sawFirst = true;
+
+      }
+
+      if (count >= cutoff) {
+        sci.summaries = new HashMap<>();
+        collector.summarize(sci);
+        collector = summarizer.collector(conf);
+        addSummary(k.getRow(), sci.summaries, count);
+        count = 0;
+      }
+
+      lastKey = k;
+    }
+
+    private List<SummaryInfo> merge(int end) {
+      List<SummaryInfo> mergedSummaries = new ArrayList<>();
+      for (int i = 0; i < end; i += 2) {
+        int mergedCount = summaries.get(i).count + summaries.get(i + 1).count;
+        summarizer.combiner(conf).merge(summaries.get(i).summary, summaries.get(i + 1).summary);
+        mergedSummaries.add(new SummaryInfo(summaries.get(i + 1).getLastRow(), summaries.get(i).summary, mergedCount));
+      }
+      return mergedSummaries;
+    }
+
+    private void addSummary(Text row, Map<String,Long> summary, int count) {
+      Preconditions.checkState(!finished);
+      summaries.add(new SummaryInfo(row, summary, count));
+
+      if (summaries.size() % 2 == 0 && summaries.size() > maxSummaries) {
+        summaries = merge(summaries.size());
+        cutoff *= 2;
+      }
+    }
+
+    boolean collapse() {
+      Preconditions.checkState(finished);
+      if (summaries.size() <= 1) {
+        return false;
+      }
+
+      int end = summaries.size();
+      if (end % 2 == 1) {
+        end--;
+      }
+
+      List<SummaryInfo> mergedSummaries = merge(end);
+
+      if (summaries.size() % 2 == 1) {
+        mergedSummaries.add(summaries.get(summaries.size() - 1));
+      }
+
+      summaries = mergedSummaries;
+
+      return true;
+    }
+
+    void finish() {
+      Preconditions.checkState(!finished);
+      // summarize last data
+      if (count > 0) {
+        sci.summaries = new HashMap<>();
+        collector.summarize(sci);
+        collector = null;
+        addSummary(lastKey.getRow(), sci.summaries, count);
+        count = 0;
+        finished = true;
+      }
+    }
+
+    public void save(DataOutputStream dos, HashMap<String,Integer> symbolTable) throws IOException {
+      Preconditions.checkState(count == 0);
+
+      dos.writeUTF(name);
+
+      if (firstRow == null) {
+        WritableUtils.writeVInt(dos, 0);
+      } else {
+        firstRow.write(dos);
+      }
+
+      // write summaries
+      WritableUtils.writeVInt(dos, summaries.size());
+      for (SummaryInfo summaryInfo : summaries) {
+        summaryInfo.getLastRow().write(dos);
+        WritableUtils.writeVInt(dos, summaryInfo.count);
+        saveSummary(dos, symbolTable, summaryInfo.summary);
+      }
+    }
+
+    private void saveSummary(DataOutputStream dos, HashMap<String,Integer> symbolTable, Map<String,Long> summary) throws IOException {
+      WritableUtils.writeVInt(dos, summary.size());
+      for (Entry<String,Long> e : summary.entrySet()) {
+        WritableUtils.writeVInt(dos, symbolTable.get(e.getKey()));
```

Comment:
It should always have a mapping.  It it does not the code will fail fast when trying to convert `null` to `int`.